internal/providers/vault/certificate.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/hashicorp/vault-client-go/schema"
+	"namespacelabs.dev/foundation/framework/secrets"
+	"namespacelabs.dev/foundation/framework/secrets/combined"
+	"namespacelabs.dev/foundation/internal/fnapi"
+	"namespacelabs.dev/foundation/internal/fnerrors"
+	"namespacelabs.dev/foundation/std/tasks"
+	"namespacelabs.dev/foundation/universe/vault"
+)
+
+const (
+	VaultJwtAudience    = "vault.namespace.systems"
+	vaultRequestTimeout = 10 * time.Second
+)
+
+func Register() {
+	p := &provider{}
+	combined.RegisterSecretsProvider(
+		func(ctx context.Context, srvRef *secrets.ServerRef, cfg *vault.Certificate) ([]byte, error) {
+			ca := cfg.GetCa()
+			if ca == nil {
+				return nil, fnerrors.BadInputError("invalid vault certificate configuration: missing CA configuration")
+			}
+
+			if err := p.Login(ctx, ca, VaultJwtAudience); err != nil {
+				return nil, err
+			}
+
+			commonName := fmt.Sprintf("%s.%s", strings.ReplaceAll(srvRef.RelPath, "/", "-"), cfg.BaseDomain)
+			return p.IssueCertificate(ctx, ca, commonName)
+		},
+	)
+}
+
+type tlsKeyPair struct {
+	PrivateKey  string   `json:"private_key"`
+	Certificate string   `json:"certificate"`
+	CaChain     []string `json:"ca_chain"`
+}
+
+type provider struct {
+	vaultClient *vaultclient.Client
+	once        sync.Once
+}
+
+func (p *provider) Login(ctx context.Context, caCfg *vault.CertificateAuthority, audience string) error {
+	var rErr error
+	p.once.Do(func() {
+		if err := tasks.Return0(ctx, tasks.Action("vault.login").Arg("namespace", caCfg.VaultNamespace).Arg("address", caCfg.VaultAddr),
+			func(ctx context.Context) error {
+				var err error
+				p.vaultClient, err = vaultclient.New(
+					vaultclient.WithAddress(caCfg.VaultAddr),
+					vaultclient.WithRequestTimeout(vaultRequestTimeout),
+				)
+				if err != nil {
+					return fnerrors.InvocationError("vault", "failed to create vault client: %w", err)
+				}
+
+				p.vaultClient.SetNamespace(caCfg.VaultNamespace)
+
+				idTokenResp, err := fnapi.IssueIdToken(ctx, audience, 1)
+				if err != nil {
+					return err
+				}
+
+				loginResp, err := p.vaultClient.Auth.JwtLogin(ctx, schema.JwtLoginRequest{Jwt: idTokenResp.IdToken},
+					vaultclient.WithMountPath(caCfg.GetAuthMethod()),
+				)
+				if err != nil {
+					return fnerrors.InvocationError("vault", "failed to login to vault: %w", err)
+				}
+
+				if loginResp.Auth == nil {
+					return fnerrors.InvocationError("vault", "missing vault login auth data: %w", err)
+				}
+
+				p.vaultClient.SetToken(loginResp.Auth.ClientToken)
+				return nil
+			},
+		); err != nil {
+			rErr = err
+		}
+	})
+
+	return rErr
+}
+
+func (p *provider) IssueCertificate(ctx context.Context, ca *vault.CertificateAuthority, cn string) ([]byte, error) {
+	return tasks.Return(ctx, tasks.Action("vault.issue-certificate").Arg("issuer", ca.GetIssuer()).Arg("common-name", cn),
+		func(ctx context.Context) ([]byte, error) {
+			pkiMount, role, ok := strings.Cut(ca.GetIssuer(), "/")
```

Comment:
That could be because strings.Cut is somewhat recent (go1.18). I like Cut better since it always returns three values, with SplitN it is easy to forget about the edge case (CA contains no "/") and index into the result which will panic.

Reply:
+1 to @attilaolah , using `strings.Cut` is safer.